### PR TITLE
Gtk+ Haskell GUI Tutorial

### PR DESCRIPTION
**Pull request by jlgalarza3**
Tutorial code for:

https://www.stackbuilders.com/blog/gui-application/

### CODE REVIEW

File: haskell/gtk-gui/gtk-gui.cabal [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+library
+  exposed-modules:
+      Lib
+  other-modules:
+      Paths_gtk_gui
+  autogen-modules:
+      Paths_gtk_gui
+  hs-source-dirs:
+      src
+  ghc-options: -Wall -Wcompat -Widentities -Wincomplete-record-updates -Wincomplete-uni-patterns -Wmissing-export-lists -Wmissing-home-modules -Wpartial-fields -Wredundant-constraints
```
Comment (felixminom) [inline haskell/gtk-gui/gtk-gui.cabal:34]: Do we need all of these enabled?

File: haskell/gtk-gui/gtk-gui.cabal [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      Paths_gtk_gui
+  hs-source-dirs:
+      app
+  ghc-options: -Wall -Wcompat -Widentities -Wincomplete-record-updates -Wincomplete-uni-patterns -Wmissing-export-lists -Wmissing-home-modules -Wpartial-fields -Wredundant-constraints -threaded -rtsopts -with-rtsopts=-N
+  build-depends:
+      base >=4.7 && <5
+    , gtk-gui
+  default-language: Haskell2010
+
+test-suite gtk-gui-test
```
Comment (felixminom) [inline haskell/gtk-gui/gtk-gui.cabal:54]: Do we have a test suite?

File: haskell/gtk-gui/gtk-gui.cabal [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      Paths_gtk_gui
+  hs-source-dirs:
+      src
+  ghc-options: -Wall -Wcompat -Widentities -Wincomplete-record-updates -Wincomplete-uni-patterns -Wmissing-export-lists -Wmissing-home-modules -Wpartial-fields -Wredundant-constraints
+  build-depends:
+      base >=4.7 && <5,
+      gtk3
+  default-language: Haskell2010
+
+executable gtk-gui-exe
```
Comment (felixminom) [inline haskell/gtk-gui/gtk-gui.cabal:40]: Probably we don't need to generate an exe.

File: haskell/gtk-gui/package.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
+name:                gtk-gui
+version:             0.1.0.0
+github:              "githubuser/gtk-gui"
+license:             BSD-3-Clause
+author:              "Author name here"
+maintainer:          "example@example.com"
```
Comment (felixminom) [inline haskell/gtk-gui/package.yaml:6]: We should update this info, the author is Mark Karpov but we can change the maintainer.

File: haskell/gtk-gui/package.yaml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+# synopsis:            Short description of your package
+# category:            Web
+
+# To avoid duplicated efforts in documentation and dealing with the
+# complications of embedding Haddock markup inside cabal files, it is
+# common to point users to the README.md file.
+description:         Please see the README on GitHub at <https://github.com/githubuser/gtk-gui#readme>
+
+dependencies:
+- base >= 4.7 && < 5
```
Comment (felixminom) [inline haskell/gtk-gui/package.yaml:23]: Probably we can define the boundaries for the gtk library

File: haskell/gtk-gui/stack.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+resolver:
+  url: https://raw.githubusercontent.com/commercialhaskell/stackage-snapshots/master/lts/22/11.yaml
```
Comment (felixminom) [inline haskell/gtk-gui/stack.yaml:2]: I'll use the `resolver: lts-21.13` instead, is more commonly used across haskell projects.

**Review by felixminom — COMMENTED** [review]
Looking good, left some comments, let me know what you think.

File: haskell/gtk-gui/gtk-gui.cabal (hunk 1 above)
Comment (jlgalarza3) [reply haskell/gtk-gui/gtk-gui.cabal]: Not really, I've removed those!

File: haskell/gtk-gui/gtk-gui.cabal (hunk 2 above)
Comment (jlgalarza3) [reply haskell/gtk-gui/gtk-gui.cabal]: Not at the moment, I have removed it

File: haskell/gtk-gui/package.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+copyright:           "2016 Mark Karpov"
+
+extra-source-files:
+- README.md
+
+description:         Please see the README on GitHub at <https://github.com/githubuser/gtk-gui#readme>
+
+dependencies:
+- base >= 4.7 && < 5
+- gtk3 >= 0.15.8
```
Comment (felixminom) [inline haskell/gtk-gui/package.yaml:16]: We should restrict the upper bounder so we don't have unexpected breaking changes.

File: haskell/gtk-gui/README.md [hunk 8]
```diff
@@ -0,0 +1 @@
+# gtk-gui
```
Comment (felixminom) [inline haskell/gtk-gui/README.md:1]: We should provide an actual README so people can build this in their machines.

**Review by felixminom — COMMENTED** [review]
Looking good just a couple of minor comments, let me know when it's ready.

**Comment by jlgalarza3** [comment]
Done! @felixminom Could you please take a second look?

File: haskell/gtk-gui/README.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
+# Creating a GUI application in Haskell with GTK+
+
+## Description
+
+This project serves as a companion codebase for Stack Builder's blog post ["Creating a GUI application in Haskell."](https://www.stackbuilders.com/blog/gui-application/)
+
+## Requirements
+
+- Stack 2.11.1
+- Cabal 3.10.2.1
```
Comment (felixminom) [inline haskell/gtk-gui/README.md:10]: we don't need this, it's enough with the stack version, that will take care of the rest.

File: haskell/gtk-gui/README.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+   ```bash
+   git clone https://github.com/stackbuilders/blog-code.git
+   cd haskell/gtk-gui/
+   ```
+
+2. **Install the necessary tools for gtk2hs**: You need to install some tools to build GTK+ bindings. Run the following commands:
+
+   ```bash
+   stack install alex happy
```
Comment (felixminom) [inline haskell/gtk-gui/README.md:29]: If we need these tools then we should make them part of the dependencies.

File: haskell/gtk-gui/README.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+3. **Build the Project**: Use Stack to build the project.
+
+   ```bash
+   stack build
+   ```
+
+4. **Run the Project**: Run the project with the following command:
+
+   ```bash
+    stack exec gtk-gui-exe
```
Comment (felixminom) [inline haskell/gtk-gui/README.md:44]: these should be the only things required to be ran.

**Review by felixminom — COMMENTED** [review]
Almost there.

**Comment by felixminom** [comment]
We should have a warning here or in the blog post that this won't work for ARM architectures. I got this error when trying to install gtk+3 on my M1 machine. 
image

**Comment by jlgalarza3** [comment]
@felixminom Thank you. Added to README and blog post

File: haskell/gtk-gui/package.yaml [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+description:         Please see the README on GitHub at <https://github.com/githubuser/gtk-gui#readme>
+
+dependencies:
+- base >= 4.7 && < 5
+- gtk3 >= 0.15.8 && < 0.15.9
+- alex
+- happy
+- haskell-gi
+- gtk2hs-buildtools
```
Comment (felixminom) [inline haskell/gtk-gui/package.yaml:20]: Should this be bounded as well? I don't see the changes in the cabal.

**Review by felixminom — COMMENTED** [review]
I think this is the last one.

**Comment by jlgalarza3** [comment]
Thanks for noticing @felixminom , apparently, it was because the .cabal file was created with another version of hpack. I just fixed it.

File: haskell/gtk-gui/package.yaml [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+github:              "stackbuilders/blog-code"
+license:             BSD-3-Clause
+author:              "Mark Karpov"
+maintainer:          "jgalarza@stackbuilders.com"
+copyright:           "2016 Mark Karpov"
+
+extra-source-files:
+- README.md
+
+description:         Please see the README on GitHub at <https://github.com/githubuser/gtk-gui#readme>
```
Comment (jpvillaisaza) [inline haskell/gtk-gui/package.yaml:12]: This description doesn't seem to have a correct link.

File: haskell/gtk-gui/stack.yaml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+resolver: lts-21.13
```
Comment (jpvillaisaza) [inline haskell/gtk-gui/stack.yaml:1]: Is there a reason for using LTS 21.13 (GHC 9.4.7) instead of 22.13 (GHC 9.6.4)? Would it work with 22?

**Review by jpvillaisaza — APPROVED** [review]

File: haskell/gtk-gui/stack.yaml (hunk 14 above)
Comment (felixminom) [reply haskell/gtk-gui/stack.yaml:1]: Didn't notice this but we can bump to the 22.13 @jlgalarza3

**Review by felixminom — APPROVED** [review]
Looking good, JP left some minor comments but after that we're good to go. Just make sure that you can build this package from scratch.

File: haskell/gtk-gui/package.yaml (hunk 13 above)
Comment (jlgalarza3) [reply haskell/gtk-gui/package.yaml]: I have removed it as I do not see it as completely necessary.

File: haskell/gtk-gui/stack.yaml (hunk 14 above)
Comment (jlgalarza3) [reply haskell/gtk-gui/stack.yaml]: DONE! It works with 22.13

**Comment by jlgalarza3** [comment]
Thank you @felixminom @jpvillaisaza. It works correctly with `lts- 22.13`. Merging this PR :rocket:

**Pull request merged**